src/tests/vos_perf.c

```diff
@@ -0,0 +1,864 @@
+		else
+			rc = vos_fetch_end(ioh, NULL, rc);
+	}
+
+	TS_TIME_END(duration, start);
+	return rc;
+}
+
+struct vos_ult_arg {
+	struct io_credit	*cred;
+	double			*duration;
+	daos_epoch_t		 epoch;
+	enum ts_op_type		 op_type;
+	int			 obj_idx;
+	int			 status;
+};
+
+static void
+vos_update_or_fetch_ult(void *arg)
+{
+	struct vos_ult_arg *ult_arg = arg;
+
+	ult_arg->status = _vos_update_or_fetch(ult_arg->obj_idx,
+					       ult_arg->op_type,
+					       ult_arg->cred,
+					       ult_arg->epoch,
+					       ult_arg->duration);
+}
+
+static int
+vos_update_or_fetch(int obj_idx, enum ts_op_type op_type,
+		    struct io_credit *cred, daos_epoch_t epoch,
+		    bool sync, double *duration)
+{
+	ABT_thread		thread;
+	struct vos_ult_arg	ult_arg;
+	int			rc;
+
+	if (!ts_in_ult)
+		return _vos_update_or_fetch(obj_idx, op_type, cred, epoch,
+					    duration);
+
+	ult_arg.op_type = op_type;
+	ult_arg.cred = cred;
+	ult_arg.epoch = epoch;
+	ult_arg.duration = duration;
+	ult_arg.obj_idx = obj_idx;
+	rc = ABT_thread_create_on_xstream(abt_xstream, vos_update_or_fetch_ult,
+					  &ult_arg, ABT_THREAD_ATTR_NULL,
+					  &thread);
+	if (rc != ABT_SUCCESS)
+		return rc;
+
+	rc = ABT_thread_join(thread);
+	if (rc != ABT_SUCCESS)
+		return rc;
+
+	ABT_thread_free(&thread);
+	rc = ult_arg.status;
+	return rc;
+}
+
+static int
+objects_query(struct pf_param *param)
+{
+	daos_epoch_t epoch = crt_hlc_get();
+	char		*akey = "0";
+	d_iov_t		dkey_iov;
+	d_iov_t		akey_iov;
+	daos_recx_t	recx;
+	int		i;
+	int		rc = 0;
+	uint64_t	start = 0;
+
+
+	d_iov_set(&akey_iov, akey, 1);
+
+	TS_TIME_START(&param->pa_duration, start);
+
+	for (i = 0; i < ts_obj_p_cont; i++) {
+		d_iov_set(&dkey_iov, NULL, 0);
+		rc = vos_obj_query_key(ts_ctx.tsc_coh, ts_uoids[i],
+				       DAOS_GET_MAX | DAOS_GET_DKEY |
+				       DAOS_GET_RECX, epoch, &dkey_iov,
+				       &akey_iov, &recx, NULL);
+		if (rc != 0 && rc != -DER_NONEXIST)
+			break;
+		if (param->pa_verbose) {
+			if (rc == -DER_NONEXIST) {
+				printf("query_key "DF_UOID ": -DER_NONEXIST\n",
+				       DP_UOID(ts_uoids[i]));
+			} else {
+				printf("query_key "DF_UOID ": dkey="DF_U64
+				       " recx="DF_RECX"\n",
+				       DP_UOID(ts_uoids[i]),
+				       *(uint64_t *)dkey_iov.iov_buf,
+				       DP_RECX(recx));
+			}
+		}
+		rc = 0;
+	}
+
+	TS_TIME_END(&param->pa_duration, start);
+
+	return rc;
+}
+
+typedef int (*iterate_cb_t)(daos_handle_t ih, vos_iter_entry_t *key_ent,
+			    vos_iter_param_t *param);
+
+static int
+iter_cb(daos_handle_t ih, vos_iter_entry_t *entry, vos_iter_type_t type, vos_iter_param_t *param,
+	void *cb_arg, unsigned int *acts)
+{
+	struct pf_param *ppa = cb_arg;
+ 
+	if (ppa->pa_verbose) {
+		switch (type) {
+		case VOS_ITER_DKEY:
+			D_PRINT("\tdkey ="DF_KEY"\n", DP_KEY(&entry->ie_key));
+ 			break;
+		case VOS_ITER_AKEY:
+			D_PRINT("\takey ="DF_KEY"\n", DP_KEY(&entry->ie_key));
+			break;
+		case VOS_ITER_SINGLE:
+			D_PRINT("\tsingv="DF_U64" bytes\n", entry->ie_rsize);
+ 			break;
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs